### ISSUE TITLE
`unnecessary_lazy_evaluations`: Shouldn't warn about values with significant `Drop` impl

### ISSUE THREAD

**Issue by sdroege** [rust-lang/rust-clippy#9427]
### Summary

The case in question that was wrongly reported is the following

```
warning: unnecessary closure used with `bool::then`
   --> glib/src/convert.rs:122:9
    |
122 |         (iconv as isize != -1).then(|| Self(iconv))
    |         ^^^^^^^^^^^^^^^^^^^^^^^--------------------
    |                                |
    |                                help: use `then_some(..)` instead: `then_some(Self(iconv))`
    |
    = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
```

`Self` here has a significant `Drop` impl, specifically it will close the `isize` there (it's basically like a `fd` and closing `-1` is UB).

### Lint Name

unnecessary_lazy_evaluations

### Reproducer

I tried this code:

```rust
struct Foo(i32);

impl Drop for Foo {
    fn drop(&mut self) {
        println!("{}", self.0);
    }
}

fn main() {
    (0 == 1).then(|| Foo(0));
}
```

I saw this happen:

```
warning: unnecessary closure used with `bool::then`
  --> src/main.rs:10:5
   |
10 |     (0 == 1).then(|| Foo(0));
   |     ^^^^^^^^^---------------
   |              |
   |              help: use `then_some(..)` instead: `then_some(Foo(0))`
   |
   = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
```

I expected to see this happen: nothing because the suggestion gives different behaviour

### Version

```text
rustc 1.65.0-nightly (289279de1 2022-09-04)
binary: rustc
commit-hash: 289279de116707f28cf9c18e4bbb8c6ec84ad75b
commit-date: 2022-09-04
host: x86_64-unknown-linux-gnu
release: 1.65.0-nightly
LLVM version: 15.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by Jules-Bertholet** [comment]
This bug is not fixed.

```rust
struct NoisyDrop;

impl Drop for NoisyDrop {
    fn drop(&mut self) {
        panic!();
    }
}

fn a() {
    let _ = false.then(|| NoisyDrop);
}
```

warns on `clippy 0.1.66 (9565dfe 2022-10-28)`

**Comment by kraktus** [comment]
can reproduce on master, will look at

**Comment by xFrednet** [comment]
#9750 should have fixed this issue :).

**Comment by ju1ius** [comment]
@xFrednet  I believe this issue is not fixed and should be reopened.

The following code:
```rs
extern crate ffi;
use core::ffi::c_char;

struct Buffer {
    ptr: *mut c_char,
    len: usize,
}
impl Buffer {
    unsafe fn from_owned_ptr(ptr: *mut c_char, len: usize) -> Option<Self> {
        (!ptr.is_null()).then(|| Self { ptr, len })
    }
}
impl Drop for Buffer {
    fn drop(&mut self) {
        unsafe { ffi::free(self.ptr) };
    }
}
```
Triggers this warning (using clippy 0.1.76):
```
warning: unnecessary closure used with `bool::then`
  --> buffer.rs:10:9
   |
48 |         (!ptr.is_null()).then(|| Self { ptr, len })
   |         ^^^^^^^^^^^^^^^^^--------------------------
   |                          |
   |                          help: use `then_some(..)` instead: `then_some(Self { ptr, len })`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
   = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default
```

**Comment by y21** [comment]
Opened #12097 for that one
